jerry-core/ecma/operations/ecma-jobqueue.h

```diff
@@ -0,0 +1,63 @@
+/* Copyright JS Foundation and other contributors, http://js.foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef ECMA_JOB_QUEUE_H
+#define ECMA_JOB_QUEUE_H
+
+#ifndef CONFIG_DISABLE_ES2015_PROMISE_BUILTIN
+
+/** \addtogroup ecma ECMA
+ * @{
+ *
+ * \addtogroup ecmajobqueue ECMA Job Queue related routines
+ * @{
+ */
+
+typedef struct ecma_job_t ecma_job_t;
+typedef void (*ecma_job_handler_t) (ecma_job_t *);
+
+/**
+ * Description of general ecma job.
+ * Usually, it is used as the header of a specific job
+ */
+struct ecma_job_t
+{
+  ecma_job_handler_t handler; /**< the handler funtion for the job */
+};
```

Comment:
Actually these are my issues with the current approach. An `ecma_job_t` is defined here, which is basically a `jerry_job_t`. Furthermore both definitions require a circular definitions, and that is never a good practice. The structures below also follows some class in C programming model.

My suggestion is: keep the callback at jerry_port.h/ The enqueue get the callback and a user pointer. Since the callback is not stored in the job, `ecma_job_promise_reaction_job_t` and `ecma_job_promise_thenable_job_t` do not need to depend on anything. We will use different callbacks for reaction and thenable (what a strange word) so there will be no confusion. I think we need less type casts overall.